src/sparseml/pytorch/image_classification/utils/trainer.py

```diff
@@ -0,0 +1,324 @@
+    :param local_rank: The local rank of the current process. Should be set
+        by torch and passed to the trainer.
+    :param rank: The ddp rank of the current process. Defaults to -1
+    :param device: The device to train on. Defaults to torch default device.
+    :param use_mixed_precision: Whether to use mixed precision FP16 training.
+        Defaults to False.
+    :param sparse_transfer_learn: True if transfer learning on a sparse model,
+        else False. Defaults to False.
+    :param val_loader: A DataLoader for validation data.
+    :param train_loader: A DataLoader for training data.
+    :param is_main_process: Whether the current process is the main process,
+        while training using DDP. Defaults to True.
+    :param loggers: A list of loggers to use during training process.
+    :param loss_fn: A Callabale loss function for training and validation
+        losses.
+    :param init_lr: The initial learning rate for the optimizer.Defaults to
+        1e-9
+    :param optim_name: str respresenting the optimizer type to use, one of
+        [SGD, Adam, RMSprop]. Defaults to `adam`.
+    :param optim_kwargs: dict of additional kwargs to pass to the optimizer.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        key: str,
+        recipe_path: str,
+        local_rank: int,
+        rank: int = -1,
+        device: str = default_device(),
+        use_mixed_precision: bool = False,
+        sparse_transfer_learn: bool = False,
+        val_loader: Optional[DataLoader] = None,
+        train_loader: Optional[DataLoader] = None,
+        is_main_process: bool = True,
+        loggers: Optional[List[Any]] = None,
+        loss_fn: Callable = lambda: CrossEntropyLossWrapper(),
+        init_lr=1e-9,
+        optim_name="adam",
+        optim_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Initializes the trainer.
+        """
+        # self.checkpoint_path = checkpoint_path
+        self.recipe_path = recipe_path
+        # self.num_classes = num_classes
+        self.local_rank = local_rank
+        self.rank = rank
+        self.is_main_process = is_main_process
+        self.optim_kwargs = optim_kwargs or {}
+        self.device = device
+        self.use_mixed_precision = use_mixed_precision
+        self.model, self.key = model, key
+        self.loss_fn = loss_fn()
+        self.init_lr = init_lr
+        self.val_loader = val_loader
+        self.train_loader = train_loader
+        self.loggers = loggers
+        self.sparse_transfer_learn = sparse_transfer_learn
+
+        self.val_loss = loss_fn()
+        _LOGGER.info(f"created loss for validation: {self.val_loss}")
+
+        self.optim_name = optim_name
+
+        # To be re-initialized on first train epoch if needed
+        self.epoch, self.optim, self.manager, self.train_loss, self.trainer = (
+            0,
+            None,
+            None,
+            None,
+            None,
+        )
```

Comment:
nit: initialize each variable on separate line. 1. much easier to read 2. fewer lines